File: sdk-tests/dynamic-swap/prover/src/pool_withdraw.rs
use std::collections::HashMap;

use zolana_gnark_ffi_prover::{decimal, utxo_proof_inputs, ProofInputMap};

use crate::{CircuitId, OrderProof, ProofInputUtxo, PROVER};

/// Proof inputs for the `pool_withdraw` circuit: 1-in (pool note) / 1-out
/// (pool change), the exact IN1_OUT1 shape. The public `amount` leaves the
/// pool through the transact's SplWithdrawal leg; the change note keeps
/// `booked_in - amount` as its data hash, rejected in-circuit if negative, so
/// a withdrawal can only consume counted value.
#[derive(Debug, Clone)]
pub struct PoolWithdrawProofInputs {
    pub public_input_hash: [u8; 32],
    pub private_tx_hash: [u8; 32],
    /// The pool_authority PDA's owner-hash (`PoolAuthorityOwnerHash`), bound
    /// to both pool notes.
    pub pool_authority_owner_hash: [u8; 32],
    /// The pair's destination-asset commitment (`DestinationAsset`).
    pub destination_asset: [u8; 32],
    /// The public withdrawn amount.
    pub amount: u64,
    pub pool_in: ProofInputUtxo,
    pub pool_out: ProofInputUtxo,
    pub external_data_hash: [u8; 32],
    pub private_tx_blinding: [u8; 32],
}

impl PoolWithdrawProofInputs {
    fn witness(&self) -> ProofInputMap {
        let mut map = HashMap::new();
        map.insert(
            "Public_PublicInputHash".to_string(),
            vec![decimal(&self.public_input_hash)],
        );
        map.insert(
            "Public_PrivateTxHash".to_string(),
            vec![decimal(&self.private_tx_hash)],
        );
        map.insert(
            "Public_PoolAuthorityOwnerHash".to_string(),
            vec![decimal(&self.pool_authority_owner_hash)],
        );
        map.insert(
            "Public_DestinationAsset".to_string(),
            vec![decimal(&self.destination_asset)],
        );
        map.insert("Public_Amount".to_string(), vec![self.amount.to_string()]);
        map.insert(
            "ExternalDataHash".to_string(),
            vec![decimal(&self.external_data_hash)],
        );
        map.insert(
            "PrivateTxBlinding".to_string(),
            vec![decimal(&self.private_tx_blinding)],
        );
        for (key, value) in utxo_proof_inputs(&self.pool_in, "PoolIn")
            .into_iter()
            .chain(utxo_proof_inputs(&self.pool_out, "PoolOut"))
        {
            map.insert(key, value);
        }
        map
    }

    pub fn prove(&self) -> zolana_gnark_ffi_prover::Result<OrderProof> {
        Ok(PROVER
            .prove(CircuitId::PoolWithdraw, &self.witness())?
            .compress()?
            .into())
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use super::*;
    use zolana_gnark_ffi_prover::utxo_proof_input_keys;

    fn sample() -> PoolWithdrawProofInputs {
        PoolWithdrawProofInputs {
            public_input_hash: [1; 32],
            private_tx_hash: [2; 32],
            pool_authority_owner_hash: [3; 32],
            destination_asset: [4; 32],
            amount: 25,
            pool_in: ProofInputUtxo::default(),
            pool_out: ProofInputUtxo::default(),
            external_data_hash: [5; 32],
            private_tx_blinding: [9; 32],
        }
    }

    #[test]
    fn witness_key_set_matches_circuit_fields() {
        let witness = sample().witness();
        let keys: HashSet<&str> = witness.keys().map(String::as_str).collect();

        let mut expected: Vec<String> = vec![
            "Public_PublicInputHash".to_string(),
            "Public_PrivateTxHash".to_string(),
            "Public_PoolAuthorityOwnerHash".to_string(),
            "Public_DestinationAsset".to_string(),
            "Public_Amount".to_string(),
            "ExternalDataHash".to_string(),
            "PrivateTxBlinding".to_string(),
        ];
        for prefix in ["PoolIn", "PoolOut"] {
            expected.extend(utxo_proof_input_keys(prefix));
        }

        let expected: HashSet<&str> = expected.iter().map(String::as_str).collect();
        assert_eq!(keys, expected);
    }
}

File: sdk-tests/dynamic-swap/prover/circuits/pool_withdraw/pool_withdraw.go
// Package pool_withdraw unshields a public amount of the destination asset
// from one pool note.
package pool_withdraw

import (
	"github.com/consensys/gnark/frontend"

	"zolana/gnarksdk"
)

// Circuit spends one pool note into one pool change note, the exact supported
// IN1_OUT1 shape. The public SPL leg is the transact's SplWithdrawal interface
// transfer (the SPP transfer circuit balances it); this circuit's job is the
// booked accounting: the change note keeps booked_in - amount, and the 64-bit
// range checks reject both an overdrawn amount and a negative booked, so a
// withdrawal can only consume counted value. The native program rejects a
// zero amount before proof verification.
type Circuit struct {
	Public PublicInputs

	PoolIn  gnarksdk.Utxo
	PoolOut gnarksdk.Utxo

	ExternalDataHash  frontend.Variable
	PrivateTxBlinding frontend.Variable
}

func (c *Circuit) Define(api frontend.API) error {
	poolInHash := c.checkPoolInputUtxo(api)
	poolOutHash := c.checkPoolChangeOutputUtxo(api)

	// 1-in/1-out, exactly the supported IN1_OUT1 shape.
	privateTxHash := gnarksdk.PrivateTxHash(
		api,
		[]frontend.Variable{poolInHash},
		[]frontend.Variable{poolOutHash},
		c.ExternalDataHash,
		c.PrivateTxBlinding,
	)
	api.AssertIsEqual(privateTxHash, c.Public.PrivateTxHash)

	c.Public.Check(api)
	return nil
}

// PublicInputs folds PrivateTxHash with the pool_authority owner-hash (the
// native program recomputes it: PDA re-derived, zero-secret nullifier pubkey),
// the pair's destination asset, and the withdrawn Amount the program pays out
// through the SplWithdrawal leg and subtracts from available_liquidity.
type PublicInputs struct {
	PublicInputHash frontend.Variable `gnark:",public"`

	PrivateTxHash frontend.Variable
	// The pool_authority PDA's owner-hash, bound to both pool notes so only
	// program-locked liquidity can fund a withdrawal and the change stays
	// locked.
	PoolAuthorityOwnerHash frontend.Variable
	// The pair's destination asset, fed on-chain from Pair.destination_asset.
	DestinationAsset frontend.Variable
	// The public withdrawn amount, fed on-chain from the instruction data (the
	// program checks it against available_liquidity and the SplWithdrawal leg).
	Amount frontend.Variable
}

func (p PublicInputs) Check(api frontend.API) {
	publicInputHash := gnarksdk.Poseidon(
		api,
		p.PrivateTxHash,
		p.PoolAuthorityOwnerHash,
		p.DestinationAsset,
		p.Amount,
	)
	api.AssertIsEqual(p.PublicInputHash, publicInputHash)
}

func (c *Circuit) checkPoolInputUtxo(api frontend.API) frontend.Variable {
	c.PoolIn.AssertDefaultRing(api)
	api.AssertIsEqual(c.PoolIn.Owner, c.Public.PoolAuthorityOwnerHash)
	api.AssertIsEqual(c.PoolIn.Asset, c.Public.DestinationAsset)
	// booked_in = the pool note's data hash; pool notes commit booked directly.
	api.ToBinary(c.PoolIn.DataHash, 64)
	return c.PoolIn.Hash(api)
}

// checkPoolChangeOutputUtxo keeps the remainder locked in the pool with the
// withdrawn value consumed from booked: amount and booked both drop by the
// public Amount, and the 64-bit range checks reject wrap-around, so a
// withdrawal can neither overdraw the note nor exceed its counted value.
func (c *Circuit) checkPoolChangeOutputUtxo(api frontend.API) frontend.Variable {
	c.PoolOut.AssertDefaultRing(api)
	api.AssertIsEqual(c.PoolOut.Owner, c.Public.PoolAuthorityOwnerHash)
	api.AssertIsEqual(c.PoolOut.Asset, c.Public.DestinationAsset)

	api.AssertIsEqual(c.PoolOut.Amount, api.Sub(c.PoolIn.Amount, c.Public.Amount))
	api.ToBinary(c.PoolOut.Amount, 64)

	api.AssertIsEqual(c.PoolOut.DataHash, api.Sub(c.PoolIn.DataHash, c.Public.Amount))
	api.ToBinary(c.PoolOut.DataHash, 64)

	return c.PoolOut.Hash(api)
}
